implant/sliver/service/service_windows.go

```diff
@@ -93,3 +102,201 @@ func RemoveService(hostname string, serviceName string) error {
 	err = service.Delete()
 	return err
 }
+
+/*
+	Currently, golang.org/x/sys/windows/svc/mgr attempts to open the service
+	manager with elevated permissions: https://github.com/golang/go/issues/51465
+	https://github.com/golang/sys/blob/master/windows/svc/mgr/mgr.go#L34
+
+	We should not need elevated privileges to list services, so we will have to
+	define a custom service manager that asks for lower rights.
+
+	Similarly, we need a custom OpenService function that does not ask for
+	elevated rights so we can get the status and config of the service.
+*/
+
+func connectToServiceManager(hostname string, permissions uint32) (*mgr.Mgr, error) {
+	var connectHost *uint16
+
+	if hostname == "localhost" {
+		/*
+			According to Win32 API docs, if the machine name passed to OpenSCManager
+			is NULL or empty, a connection will be opened to the local machine
+			https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-openscmanagera
+		*/
+		connectHost = nil
+	} else {
+		connectHost = windows.StringToUTF16Ptr(hostname)
+	}
+	handle, err := windows.OpenSCManager(connectHost, nil, permissions)
+	if err != nil {
+		return nil, err
+	}
+	return &mgr.Mgr{Handle: handle}, nil
+}
+
+func openService(svcManager *mgr.Mgr, serviceName string, permissions uint32) (*mgr.Service, error) {
+	handle, err := windows.OpenService(svcManager.Handle, windows.StringToUTF16Ptr(serviceName), permissions)
+	if err != nil {
+		return nil, err
+	}
+	return &mgr.Service{Name: serviceName, Handle: handle}, nil
+}
+
+func buildServiceDetail(serviceName string, config mgr.Config) *sliverpb.ServiceDetails {
+	detail := &sliverpb.ServiceDetails{Name: serviceName}
+
+	detail.DisplayName = config.DisplayName
+	detail.Description = config.Description
+	switch config.StartType {
+	case mgr.StartManual:
+		detail.StartupType = "Manual"
+	case mgr.StartAutomatic:
+		detail.StartupType = "Automatic"
+	case mgr.StartDisabled:
+		detail.StartupType = "Disabled"
+	}
```

Comment:
We can probably pass integers back to the server and let the client do the translation. Having hardcoded strings in the implant isn't ideal, let's avoid it when/if we can. The constants are here: https://pkg.go.dev/golang.org/x/sys@v0.16.0/windows#SERVICE_DEMAND_START

Reply:
Good call. I will move the strings from the implant to the client.